File: clients/beebium-typescript-client/src/tube-ula.ts
/**
 * Tube ULA device inspection for the Beebium TypeScript client.
 */

import type { DeviceInspectionClient } from "./generated/debugger.js";
import type {
    TubeState as ProtoTubeState,
    TubeControlFlags as ProtoTubeControlFlags,
    TubeLatchState as ProtoTubeLatchState,
    TubeFifo24State as ProtoTubeFifo24State,
    TubeReg3State as ProtoTubeReg3State,
    TubeHostStatus as ProtoTubeHostStatus,
    TubeParasiteStatus as ProtoTubeParasiteStatus,
    TubeInterrupts as ProtoTubeInterrupts,
    TubeTransferCounters as ProtoTubeTransferCounters,
    GetTubeStateRequest,
} from "./generated/debugger.js";
import { promisify } from "./call-utils.js";

// -- Interfaces ---------------------------------------------------------------

export interface TubeControlFlags {
    /** Bit 0: Enable HIRQ from R4 */
    q: boolean;
    /** Bit 1: Enable PIRQ from R1 */
    i: boolean;
    /** Bit 2: Enable PIRQ from R4 */
    j: boolean;
    /** Bit 3: Enable PNMI from R3 */
    m: boolean;
    /** Bit 4: Two-byte mode for R3 */
    v: boolean;
    /** Bit 5: Parasite reset */
    p: boolean;
}

export interface TubeLatchState {
    /** Current data byte in the latch */
    value: number;
    /** Latch contains unread data */
    dataAvailable: boolean;
}

export interface TubeFifo24State {
    /** Number of bytes currently in the FIFO (0-24) */
    count: number;
    /** FIFO contents in read order (head to tail) */
    data: Uint8Array;
    /** count > 0 */
    dataAvailable: boolean;
}

export interface TubeReg3State {
    /** Number of bytes in the register (0-2) */
    count: number;
    /** Register contents in read order (up to 2 bytes) */
    data: Uint8Array;
    /** Complete transfer awaiting consumption */
    pending: boolean;
    /** Current threshold (1 or 2, from V flag) */
    threshold: number;
}

export interface TubeHostStatus {
    /** Offset 0: control flags + R1 status bits */
    r1Status: number;
    /** Offset 2: R2 status bits */
    r2Status: number;
    /** Offset 4: R3 status bits */
    r3Status: number;
    /** Offset 6: R4 status bits */
    r4Status: number;
}

export interface TubeParasiteStatus {
    /** Offset 0: control flags + R1 status bits */
    r1Status: number;
    /** Offset 2: R2 status bits */
    r2Status: number;
    /** Offset 4: R3 status bits (includes N flag at bit 5) */
    r3Status: number;
    /** Offset 6: R4 status bits */
    r4Status: number;
}

export interface TubeInterrupts {
    /** Host IRQ: Q=1 AND R4 P-to-H has data */
    hirq: boolean;
    /** Parasite IRQ: (I=1 AND R1 H-to-P has data) OR (J=1 AND R4 H-to-P has data) */
    pirq: boolean;
    /** PNMI level (combinational, before edge detection) */
    pnmiLevel: boolean;
    /** PNMI latched rising edge */
    pnmiEdge: boolean;
}

export interface TubeTransferCounters {
    /** Host-to-parasite direction */
    r1H2pWrites: number;
    r1H2pReads: number;
    r2H2pWrites: number;
    r2H2pReads: number;
    r3H2pWrites: number;
    r3H2pReads: number;
    r4H2pWrites: number;
    r4H2pReads: number;
    /** Parasite-to-host direction */
    r1P2hWrites: number;
    r1P2hReads: number;
    r2P2hWrites: number;
    r2P2hReads: number;
    r3P2hWrites: number;
    r3P2hReads: number;
    r4P2hWrites: number;
    r4P2hReads: number;
}

export interface TubeUlaState {
    /** Whether the Tube socket is populated */
    enabled: boolean;
    /** Control flags (6 bits, written by host) */
    controlFlags: TubeControlFlags;
    /** Register 1: H-to-P is a 1-byte latch */
    r1H2p: TubeLatchState;
    /** Register 1: P-to-H is a 24-byte FIFO */
    r1P2h: TubeFifo24State;
    /** Register 2: 1-byte latch in each direction */
    r2H2p: TubeLatchState;
    r2P2h: TubeLatchState;
    /** Register 3: 2-byte shift register in each direction */
    r3H2p: TubeReg3State;
    r3P2h: TubeReg3State;
    /** Register 4: 1-byte latch in each direction */
    r4H2p: TubeLatchState;
    r4P2h: TubeLatchState;
    /** Status registers (as read by host) */
    hostStatus: TubeHostStatus;
    /** Status registers (as read by parasite) */
    parasiteStatus: TubeParasiteStatus;
    /** Interrupt outputs */
    interrupts: TubeInterrupts;
    /** Bus stretching state */
    hostStretched: boolean;
    /** Per-register transfer counters */
    counters: TubeTransferCounters;
}

// -- Formatting helpers -------------------------------------------------------

function formatControlFlags(f: TubeControlFlags): string {
    const raw =
        (f.q ? 1 : 0) |
        (f.i ? 2 : 0) |
        (f.j ? 4 : 0) |
        (f.m ? 8 : 0) |
        (f.v ? 16 : 0) |
        (f.p ? 32 : 0);
    const flags: string[] = [];
    if (f.q) flags.push("Q");
    if (f.i) flags.push("I");
    if (f.j) flags.push("J");
    if (f.m) flags.push("M");
    if (f.v) flags.push("V");
    if (f.p) flags.push("P");
    return `Flags=${raw.toString(16).toUpperCase().padStart(2, "0")} [${flags.join(" ") || "none"}]`;
}

function formatLatch(l: TubeLatchState): string {
    const avail = l.dataAvailable ? "AVAIL" : "empty";
    return `$${l.value.toString(16).toUpperCase().padStart(2, "0")} (${avail})`;
}

function formatFifo24(f: TubeFifo24State): string {
    const avail = f.dataAvailable ? "AVAIL" : "empty";
    const bytes = Array.from(f.data.slice(0, 8));
    let hexData = bytes.map(b => b.toString(16).toUpperCase().padStart(2, "0")).join(" ");
    if (f.data.length > 8) {
        hexData += " ...";
    }
    return `count=${f.count}/24 (${avail}) [${hexData}]`;
}

function formatReg3(r: TubeReg3State): string {
    const mode = r.threshold === 2 ? "2-byte" : "1-byte";
    const pend = r.pending ? "PENDING" : "idle";
    const hexData = Array.from(r.data).map(b => b.toString(16).toUpperCase().padStart(2, "0")).join(" ");
    return `count=${r.count}/${r.threshold} (${pend}, ${mode}) [${hexData}]`;
}

function formatStatus(label: string, s: TubeHostStatus | TubeParasiteStatus): string {
    const hex2 = (n: number) => n.toString(16).toUpperCase().padStart(2, "0");
    return `${label}: R1=$${hex2(s.r1Status)} R2=$${hex2(s.r2Status)} R3=$${hex2(s.r3Status)} R4=$${hex2(s.r4Status)}`;
}

function formatInterrupts(intr: TubeInterrupts): string {
    const active: string[] = [];
    if (intr.hirq) active.push("HIRQ");
    if (intr.pirq) active.push("PIRQ");
    if (intr.pnmiLevel) active.push("PNMI-level");
    if (intr.pnmiEdge) active.push("PNMI-edge");
    return active.join(" ") || "none";
}

function formatCounters(c: TubeTransferCounters): string {
    const lines = ["Transfer counters:"];
    const regs = [
        { reg: 1, h2pW: c.r1H2pWrites, h2pR: c.r1H2pReads, p2hW: c.r1P2hWrites, p2hR: c.r1P2hReads },
        { reg: 2, h2pW: c.r2H2pWrites, h2pR: c.r2H2pReads, p2hW: c.r2P2hWrites, p2hR: c.r2P2hReads },
        { reg: 3, h2pW: c.r3H2pWrites, h2pR: c.r3H2pReads, p2hW: c.r3P2hWrites, p2hR: c.r3P2hReads },
        { reg: 4, h2pW: c.r4H2pWrites, h2pR: c.r4H2pReads, p2hW: c.r4P2hWrites, p2hR: c.r4P2hReads },
    ];
    for (const r of regs) {
        const h2pDelta = r.h2pW - r.h2pR;
        const p2hDelta = r.p2hW - r.p2hR;
        lines.push(
            `  R${r.reg}: H->P w=${r.h2pW} r=${r.h2pR} (delta=${h2pDelta})` +
            `  P->H w=${r.p2hW} r=${r.p2hR} (delta=${p2hDelta})`
        );
    }
    return lines.join("\n");
}

/** Format a complete Tube ULA state snapshot for display. */
export function formatTubeUlaState(s: TubeUlaState): string {
    if (!s.enabled) {
        return "Tube: not enabled";
    }
    const lines = [
        `Tube ULA: ${formatControlFlags(s.controlFlags)}`,
        `  R1 H->P: ${formatLatch(s.r1H2p)}`,
        `  R1 P->H: ${formatFifo24(s.r1P2h)}`,
        `  R2 H->P: ${formatLatch(s.r2H2p)}`,
        `  R2 P->H: ${formatLatch(s.r2P2h)}`,
        `  R3 H->P: ${formatReg3(s.r3H2p)}`,
        `  R3 P->H: ${formatReg3(s.r3P2h)}`,
        `  R4 H->P: ${formatLatch(s.r4H2p)}`,
        `  R4 P->H: ${formatLatch(s.r4P2h)}`,
        `  ${formatStatus("Host status    ", s.hostStatus)}`,
        `  ${formatStatus("Parasite status", s.parasiteStatus)}`,
        `  Interrupts: ${formatInterrupts(s.interrupts)}`,
    ];
    if (s.hostStretched) {
        lines.push("  Host: STRETCHED");
    }
    lines.push(`  ${formatCounters(s.counters)}`);
    return lines.join("\n");
}

// -- Conversion ---------------------------------------------------------------

const DEFAULT_CONTROL_FLAGS: TubeControlFlags = {
    q: false, i: false, j: false, m: false, v: false, p: false,
};

const DEFAULT_LATCH: TubeLatchState = { value: 0, dataAvailable: false };

const DEFAULT_FIFO24: TubeFifo24State = { count: 0, data: new Uint8Array(0), dataAvailable: false };

const DEFAULT_REG3: TubeReg3State = { count: 0, data: new Uint8Array(0), pending: false, threshold: 1 };

const DEFAULT_HOST_STATUS: TubeHostStatus = { r1Status: 0, r2Status: 0, r3Status: 0, r4Status: 0 };

const DEFAULT_PARASITE_STATUS: TubeParasiteStatus = { r1Status: 0, r2Status: 0, r3Status: 0, r4Status: 0 };

const DEFAULT_INTERRUPTS: TubeInterrupts = { hirq: false, pirq: false, pnmiLevel: false, pnmiEdge: false };

const DEFAULT_COUNTERS: TubeTransferCounters = {
    r1H2pWrites: 0, r1H2pReads: 0,
    r2H2pWrites: 0, r2H2pReads: 0,
    r3H2pWrites: 0, r3H2pReads: 0,
    r4H2pWrites: 0, r4H2pReads: 0,
    r1P2hWrites: 0, r1P2hReads: 0,
    r2P2hWrites: 0, r2P2hReads: 0,
    r3P2hWrites: 0, r3P2hReads: 0,
    r4P2hWrites: 0, r4P2hReads: 0,
};

function toControlFlags(proto: ProtoTubeControlFlags | undefined): TubeControlFlags {
    if (!proto) return { ...DEFAULT_CONTROL_FLAGS };
    return { q: proto.q, i: proto.i, j: proto.j, m: proto.m, v: proto.v, p: proto.p };
}

function toLatchState(proto: ProtoTubeLatchState | undefined): TubeLatchState {
    if (!proto) return { ...DEFAULT_LATCH };
    return { value: proto.value, dataAvailable: proto.dataAvailable };
}

function toFifo24State(proto: ProtoTubeFifo24State | undefined): TubeFifo24State {
    if (!proto) return { ...DEFAULT_FIFO24 };
    return {
        count: proto.count,
        data: proto.data instanceof Uint8Array ? proto.data : new Uint8Array(proto.data),
        dataAvailable: proto.dataAvailable,
    };
}

function toReg3State(proto: ProtoTubeReg3State | undefined): TubeReg3State {
    if (!proto) return { ...DEFAULT_REG3 };
    return {
        count: proto.count,
        data: proto.data instanceof Uint8Array ? proto.data : new Uint8Array(proto.data),
        pending: proto.pending,
        threshold: proto.threshold,
    };
}

function toHostStatus(proto: ProtoTubeHostStatus | undefined): TubeHostStatus {
    if (!proto) return { ...DEFAULT_HOST_STATUS };
    return {
        r1Status: proto.r1Status,
        r2Status: proto.r2Status,
        r3Status: proto.r3Status,
        r4Status: proto.r4Status,
    };
}

function toParasiteStatus(proto: ProtoTubeParasiteStatus | undefined): TubeParasiteStatus {
    if (!proto) return { ...DEFAULT_PARASITE_STATUS };
    return {
        r1Status: proto.r1Status,
        r2Status: proto.r2Status,
        r3Status: proto.r3Status,
        r4Status: proto.r4Status,
    };
}

function toInterrupts(proto: ProtoTubeInterrupts | undefined): TubeInterrupts {
    if (!proto) return { ...DEFAULT_INTERRUPTS };
    return {
        hirq: proto.hirq,
        pirq: proto.pirq,
        pnmiLevel: proto.pnmiLevel,
        pnmiEdge: proto.pnmiEdge,
    };
}

function toTransferCounters(proto: ProtoTubeTransferCounters | undefined): TubeTransferCounters {
    if (!proto) return { ...DEFAULT_COUNTERS };
    return {
        r1H2pWrites: proto.r1H2pWrites,
        r1H2pReads: proto.r1H2pReads,
        r2H2pWrites: proto.r2H2pWrites,
        r2H2pReads: proto.r2H2pReads,
        r3H2pWrites: proto.r3H2pWrites,
        r3H2pReads: proto.r3H2pReads,
        r4H2pWrites: proto.r4H2pWrites,
        r4H2pReads: proto.r4H2pReads,
        r1P2hWrites: proto.r1P2hWrites,
        r1P2hReads: proto.r1P2hReads,
        r2P2hWrites: proto.r2P2hWrites,
        r2P2hReads: proto.r2P2hReads,
        r3P2hWrites: proto.r3P2hWrites,
        r3P2hReads: proto.r3P2hReads,
        r4P2hWrites: proto.r4P2hWrites,
        r4P2hReads: proto.r4P2hReads,
    };
}

function toTubeUlaState(proto: ProtoTubeState): TubeUlaState {
    return {
        enabled: proto.enabled,
        controlFlags: toControlFlags(proto.controlFlags),
        r1H2p: toLatchState(proto.r1H2p),
        r1P2h: toFifo24State(proto.r1P2h),
        r2H2p: toLatchState(proto.r2H2p),
        r2P2h: toLatchState(proto.r2P2h),
        r3H2p: toReg3State(proto.r3H2p),
        r3P2h: toReg3State(proto.r3P2h),
        r4H2p: toLatchState(proto.r4H2p),
        r4P2h: toLatchState(proto.r4P2h),
        hostStatus: toHostStatus(proto.hostStatus),
        parasiteStatus: toParasiteStatus(proto.parasiteStatus),
        interrupts: toInterrupts(proto.interrupts),
        hostStretched: proto.hostStretched,
        counters: toTransferCounters(proto.counters),
    };
}

// -- Class --------------------------------------------------------------------

/**
 * Tube ULA device inspection.
 *
 * Provides side-effect-free access to Tube ULA register state for debugging.
 */
export class TubeUlaInspection {
    private readonly stub: DeviceInspectionClient;

    constructor(stub: DeviceInspectionClient) {
        this.stub = stub;
    }

    /** Get the current Tube ULA state. */
    async getState(): Promise<TubeUlaState> {
        const response = await promisify<GetTubeStateRequest, ProtoTubeState>(
            this.stub as unknown as Record<string, Function>,
            "getTubeState",
            {},
        );
        return toTubeUlaState(response);
    }
}

File: clients/beebium-python-client/src/beebium/client/tube_ula.py
# Copyright 2026 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Tube ULA device inspection for the beebium client."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from beebium.client._proto import debugger_pb2

if TYPE_CHECKING:
    from beebium.client._proto import debugger_pb2_grpc


@dataclass
class TubeControlFlags:
    """Tube ULA control flags (6 bits, written by host)."""

    q: bool = False  # Bit 0: Enable HIRQ from R4
    i: bool = False  # Bit 1: Enable PIRQ from R1
    j: bool = False  # Bit 2: Enable PIRQ from R4
    m: bool = False  # Bit 3: Enable PNMI from R3
    v: bool = False  # Bit 4: Two-byte mode for R3
    p: bool = False  # Bit 5: Coprocessor reset

    @property
    def raw(self) -> int:
        """Raw 6-bit value."""
        return (self.q << 0) | (self.i << 1) | (self.j << 2) | (self.m << 3) | (self.v << 4) | (self.p << 5)

    def __str__(self) -> str:
        flags = []
        if self.q:
            flags.append("Q")
        if self.i:
            flags.append("I")
        if self.j:
            flags.append("J")
        if self.m:
            flags.append("M")
        if self.v:
            flags.append("V")
        if self.p:
            flags.append("P")
        return f"Flags={self.raw:02X} [{' '.join(flags) or 'none'}]"


@dataclass
class TubeLatchState:
    """Single-byte latch state."""

    value: int = 0
    data_available: bool = False

    def __str__(self) -> str:
        avail = "AVAIL" if self.data_available else "empty"
        return f"${self.value:02X} ({avail})"


@dataclass
class TubeFifo24State:
    """24-byte FIFO state (R1 P-to-H)."""

    count: int = 0
    data: bytes = b""
    data_available: bool = False

    def __str__(self) -> str:
        avail = "AVAIL" if self.data_available else "empty"
        hex_data = " ".join(f"{b:02X}" for b in self.data[:8])
        if len(self.data) > 8:
            hex_data += " ..."
        return f"count={self.count}/24 ({avail}) [{hex_data}]"


@dataclass
class TubeReg3State:
    """2-byte shift register state (R3)."""

    count: int = 0
    data: bytes = b""
    pending: bool = False
    threshold: int = 1

    def __str__(self) -> str:
        mode = "2-byte" if self.threshold == 2 else "1-byte"
        pend = "PENDING" if self.pending else "idle"
        hex_data = " ".join(f"{b:02X}" for b in self.data)
        return f"count={self.count}/{self.threshold} ({pend}, {mode}) [{hex_data}]"


@dataclass
class TubeHostStatus:
    """Host-perspective status registers."""

    r1_status: int = 0
    r2_status: int = 0
    r3_status: int = 0
    r4_status: int = 0

    def __str__(self) -> str:
        return f"R1=${self.r1_status:02X} R2=${self.r2_status:02X} R3=${self.r3_status:02X} R4=${self.r4_status:02X}"


@dataclass
class TubeCoprocessorStatus:
    """Coprocessor-perspective status registers."""

    r1_status: int = 0
    r2_status: int = 0
    r3_status: int = 0
    r4_status: int = 0

    def __str__(self) -> str:
        return f"R1=${self.r1_status:02X} R2=${self.r2_status:02X} R3=${self.r3_status:02X} R4=${self.r4_status:02X}"


@dataclass
class TubeInterrupts:
    """Tube interrupt output state."""

    hirq: bool = False
    pirq: bool = False
    pnmi_level: bool = False
    pnmi_edge: bool = False

    def __str__(self) -> str:
        active = []
        if self.hirq:
            active.append("HIRQ")
        if self.pirq:
            active.append("PIRQ")
        if self.pnmi_level:
            active.append("PNMI-level")
        if self.pnmi_edge:
            active.append("PNMI-edge")
        return " ".join(active) or "none"


@dataclass
class TubeTransferCounters:
    """Per-register byte transfer counters for Tube diagnostics."""

    r1_h2p_writes: int = 0
    r1_h2p_reads: int = 0
    r2_h2p_writes: int = 0
    r2_h2p_reads: int = 0
    r3_h2p_writes: int = 0
    r3_h2p_reads: int = 0
    r4_h2p_writes: int = 0
    r4_h2p_reads: int = 0

    r1_p2h_writes: int = 0
    r1_p2h_reads: int = 0
    r2_p2h_writes: int = 0
    r2_p2h_reads: int = 0
    r3_p2h_writes: int = 0
    r3_p2h_reads: int = 0
    r4_p2h_writes: int = 0
    r4_p2h_reads: int = 0

    def __str__(self) -> str:
        lines = ["Transfer counters:"]
        for reg in range(1, 5):
            h2p_w = getattr(self, f"r{reg}_h2p_writes")
            h2p_r = getattr(self, f"r{reg}_h2p_reads")
            p2h_w = getattr(self, f"r{reg}_p2h_writes")
            p2h_r = getattr(self, f"r{reg}_p2h_reads")
            h2p_delta = h2p_w - h2p_r
            p2h_delta = p2h_w - p2h_r
            lines.append(
                f"  R{reg}: H->P w={h2p_w} r={h2p_r} (delta={h2p_delta})  P->H w={p2h_w} r={p2h_r} (delta={p2h_delta})"
            )
        return "\n".join(lines)


@dataclass
class TubeUlaState:
    """Complete Tube ULA state snapshot."""

    enabled: bool = False
    control_flags: TubeControlFlags = field(default_factory=TubeControlFlags)

    r1_h2p: TubeLatchState = field(default_factory=TubeLatchState)
    r1_p2h: TubeFifo24State = field(default_factory=TubeFifo24State)
    r2_h2p: TubeLatchState = field(default_factory=TubeLatchState)
    r2_p2h: TubeLatchState = field(default_factory=TubeLatchState)
    r3_h2p: TubeReg3State = field(default_factory=TubeReg3State)
    r3_p2h: TubeReg3State = field(default_factory=TubeReg3State)
    r4_h2p: TubeLatchState = field(default_factory=TubeLatchState)
    r4_p2h: TubeLatchState = field(default_factory=TubeLatchState)

    host_status: TubeHostStatus = field(default_factory=TubeHostStatus)
    coprocessor_status: TubeCoprocessorStatus = field(default_factory=TubeCoprocessorStatus)
    interrupts: TubeInterrupts = field(default_factory=TubeInterrupts)

    host_stretched: bool = False
    counters: TubeTransferCounters = field(default_factory=TubeTransferCounters)

    def __str__(self) -> str:
        if not self.enabled:
            return "Tube: not enabled"
        lines = [
            f"Tube ULA: {self.control_flags}",
            f"  R1 H->P: {self.r1_h2p}",
            f"  R1 P->H: {self.r1_p2h}",
            f"  R2 H->P: {self.r2_h2p}",
            f"  R2 P->H: {self.r2_p2h}",
            f"  R3 H->P: {self.r3_h2p}",
            f"  R3 P->H: {self.r3_p2h}",
            f"  R4 H->P: {self.r4_h2p}",
            f"  R4 P->H: {self.r4_p2h}",
            f"  Host status:     {self.host_status}",
            f"  Coprocessor status: {self.coprocessor_status}",
            f"  Interrupts: {self.interrupts}",
        ]
        if self.host_stretched:
            lines.append("  Host: STRETCHED")
        lines.append(f"  {self.counters}")
        return "\n".join(lines)


class TubeUlaInspection:
    """Tube ULA device inspection.

    Provides side-effect-free access to Tube ULA register state for debugging.

    Usage:
        state = bbc.tube_ula.state
        print(state)  # Full formatted dump
        print(f"R1 H->P: {state.r1_h2p}")
        if state.interrupts.hirq:
            print("Host IRQ active")
    """

    def __init__(self, stub: debugger_pb2_grpc.DeviceInspectionStub):
        self._stub = stub

    @property
    def state(self) -> TubeUlaState:
        """Get the current Tube ULA state."""
        request = debugger_pb2.GetTubeStateRequest()
        response = self._stub.GetTubeState(request)
        return _from_proto(response)


def _from_proto(pb: debugger_pb2.TubeState) -> TubeUlaState:
    """Convert a protobuf TubeState to a TubeUlaState dataclass."""
    cf = pb.control_flags
    return TubeUlaState(
        enabled=pb.enabled,
        control_flags=TubeControlFlags(
            q=cf.q,
            i=cf.i,
            j=cf.j,
            m=cf.m,
            v=cf.v,
            p=cf.p,
        ),
        r1_h2p=TubeLatchState(
            value=pb.r1_h2p.value,
            data_available=pb.r1_h2p.data_available,
        ),
        r1_p2h=TubeFifo24State(
            count=pb.r1_p2h.count,
            data=pb.r1_p2h.data,
            data_available=pb.r1_p2h.data_available,
        ),
        r2_h2p=TubeLatchState(
            value=pb.r2_h2p.value,
            data_available=pb.r2_h2p.data_available,
        ),
        r2_p2h=TubeLatchState(
            value=pb.r2_p2h.value,
            data_available=pb.r2_p2h.data_available,
        ),
        r3_h2p=TubeReg3State(
            count=pb.r3_h2p.count,
            data=pb.r3_h2p.data,
            pending=pb.r3_h2p.pending,
            threshold=pb.r3_h2p.threshold,
        ),
        r3_p2h=TubeReg3State(
            count=pb.r3_p2h.count,
            data=pb.r3_p2h.data,
            pending=pb.r3_p2h.pending,
            threshold=pb.r3_p2h.threshold,
        ),
        r4_h2p=TubeLatchState(
            value=pb.r4_h2p.value,
            data_available=pb.r4_h2p.data_available,
        ),
        r4_p2h=TubeLatchState(
            value=pb.r4_p2h.value,
            data_available=pb.r4_p2h.data_available,
        ),
        host_status=TubeHostStatus(
            r1_status=pb.host_status.r1_status,
            r2_status=pb.host_status.r2_status,
            r3_status=pb.host_status.r3_status,
            r4_status=pb.host_status.r4_status,
        ),
        coprocessor_status=TubeCoprocessorStatus(
            r1_status=pb.coprocessor_status.r1_status,
            r2_status=pb.coprocessor_status.r2_status,
            r3_status=pb.coprocessor_status.r3_status,
            r4_status=pb.coprocessor_status.r4_status,
        ),
        interrupts=TubeInterrupts(
            hirq=pb.interrupts.hirq,
            pirq=pb.interrupts.pirq,
            pnmi_level=pb.interrupts.pnmi_level,
            pnmi_edge=pb.interrupts.pnmi_edge,
        ),
        host_stretched=pb.host_stretched,
        counters=TubeTransferCounters(
            r1_h2p_writes=pb.counters.r1_h2p_writes,
            r1_h2p_reads=pb.counters.r1_h2p_reads,
            r2_h2p_writes=pb.counters.r2_h2p_writes,
            r2_h2p_reads=pb.counters.r2_h2p_reads,
            r3_h2p_writes=pb.counters.r3_h2p_writes,
            r3_h2p_reads=pb.counters.r3_h2p_reads,
            r4_h2p_writes=pb.counters.r4_h2p_writes,
            r4_h2p_reads=pb.counters.r4_h2p_reads,
            r1_p2h_writes=pb.counters.r1_p2h_writes,
            r1_p2h_reads=pb.counters.r1_p2h_reads,
            r2_p2h_writes=pb.counters.r2_p2h_writes,
            r2_p2h_reads=pb.counters.r2_p2h_reads,
            r3_p2h_writes=pb.counters.r3_p2h_writes,
            r3_p2h_reads=pb.counters.r3_p2h_reads,
            r4_p2h_writes=pb.counters.r4_p2h_writes,
            r4_p2h_reads=pb.counters.r4_p2h_reads,
        ),
    )
